docs/source/examples.rst

```diff
@@ -0,0 +1,136 @@
+.. _examples:
+
+Examples
+========
+
+Create resource for the puppet handler
+--------------------------------------
+
+Let's create an example :ref:`resource-term` for the puppet :ref:`res-handler-term`
+version 1 [#]_. The resource should install and configure OpenStack Nova API
+service.
+
+.. [#] There is also puppet handler version 2 but it is out of the scope
+   of this example.
+
+Step 1: Find an apropriate puppet module
+++++++++++++++++++++++++++++++++++++++++
+
+The `Puppet OpenStack <https://wiki.openstack.org/wiki/Puppet>`_
+module for `Nova <https://github.com/openstack/puppet-nova>`_
+provides all of the required functionality.
+
+Step 2: Define granularity scope for a resource
++++++++++++++++++++++++++++++++++++++++++++++++
+
+One may want to implement resources as atomic entities doing each its only
+task, like running one and only puppet manifest [#]_. Or as a single entity doing
+all required tasks instead. In order to configure and run the Nova API service
+at least two manifests should be executed: the
+`init.pp <https://github.com/openstack/puppet-nova/blob/master/manifests/init.pp>`_
+and the `api.pp <https://github.com/openstack/puppet-nova/blob/master/manifests/api.pp>`_ [#]_.
+
+.. [#] Puppet manifests may contain references to externally defined classess
+   or services in the catalog. Keep that in mind then designing the resource.
+
+.. [#] This assumes configuring DB and messaging entities like user, password
+   database, vhost, access rights are left out of the scope of this example.
+
+Assuming the atomic tasks approach, the example resource for Nova API service
+should only use the `api.pp` manifest. Note that the puppet handler is normally
+executed in its own isolated puppet catalog with its specific hiera data only.
+This assumes every puppet manifest called by every action to be executed as a
+separate puppet run and shares nothing.
+
+Step 3: Define resource inputs
+++++++++++++++++++++++++++++++
+
+Once the granularity scope of the resource was clearly defined, one should
+define the resource's :ref:`res-input-term` data. The puppet class `nova::api`
+contains a lots of parameters. It looks reasonable to use them as the resource
+inputs as is.
+
+.. note ::
+  There is a `helper script <https://github.com/bogdando/convert_puppet_parameters>`_
+  to convert a puppet class parameters into the format expeted by the
+  `meta.yaml` inputs file.
+
+Step 4: Implement the basic action run
+++++++++++++++++++++++++++++++++++++++
+
+Each resource should have all of the mandatory actions defined. In this example
+we define only the :ref:`ref-action-term` `run`. For the example Nova API
+resource, the action run should:
+
+- fetch the resource inputs from the hiera [#]_ ::
+
+      $resource = hiera($::resource_name)
+      $ensure_package = $resource['input']['ensure_package']
```

Comment:
can we assume that we will stick with variant implemented in puppetv2?
and 'ensure_package' will be available as 

  hiera('ensure_package', false)